samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemProducer.scala

```diff
@@ -21,134 +21,100 @@ package org.apache.samza.system.kafka
 
-        sources.asScala.foreach {p =>
-          if (p._2.exceptionInCallback.get() == null) {
-            flush(p._1)
-          }
+        // Log error for debugging purposes.
+        val exception = fatalException.get()
+        if (exception != null) {
+          error(KafkaSystemProducer.UnhandledSendErrorMsg, exception)
         }
+      } catch {
+        case e: Exception => error(KafkaSystemProducer.ErrorClosingMsg + systemName, e)
+      } finally {
+        producer = null
       }
-    } catch {
-      case e: Exception => error(e.getMessage, e)
     }
   }
 
   def register(source: String) {
-    if(sources.putIfAbsent(source, new SourceData) != null) {
-      throw new SamzaException("%s is already registered with the %s system producer" format (source, systemName))
-    }
-  }
-
-  def closeAndNullifyCurrentProducer(currentProducer: Producer[Array[Byte], Array[Byte]]) {
-    try {
-      // TODO: we should use timeout close() to make sure we fail all waiting messages in kafka 0.9+
-      currentProducer.close()
-    } catch {
-      case e: Exception => error("producer close failed", e)
-    }
-    producerLock.synchronized {
-      if (currentProducer == producer) {
-        // only nullify the member producer if it is still the same object, no point nullifying new producer
-        producer = null
-      }
-    }
   }
 
   def send(source: String, envelope: OutgoingMessageEnvelope) {
-    trace("Enqueuing message: %s, %s." format (source, envelope))
+    trace(KafkaSystemProducer.SendMsgFormat format (source, envelope))
 
     val topicName = envelope.getSystemStream.getStream
     if (topicName == null || topicName == "") {
-      throw new IllegalArgumentException(StreamNameNullOrEmptyErrorMsg)
+      throw new IllegalArgumentException(KafkaSystemProducer.StreamNameNullOrEmptyErrorMsg)
     }
 
-    val sourceData = sources.get(source)
-    if (sourceData == null) {
-      throw new IllegalArgumentException("Source %s must be registered first before send." format source)
-    }
-
-    val exception = sourceData.exceptionInCallback.getAndSet(null)
-    if (exception != null) {
+    val globalProducerException = fatalException.get()
+    if (globalProducerException != null) {
       metrics.sendFailed.inc
-      throw exception  // in case the caller catches all exceptions and will try again
-    }
-
-    // lazy initialization of the producer
-    if (producer == null) {
-      producerLock.synchronized {
-        if (producer == null) {
-          info("Creating a new producer for system %s." format systemName)
-          producer = getProducer()
-        }
-      }
+      throw new SystemProducerException(KafkaSystemProducer.PreviousErrorMsg, globalProducerException)
     }
 
     val currentProducer = producer
     if (currentProducer == null) {
-      throw new SamzaException("Kafka system producer is not available.")
+      throw new SystemProducerException(KafkaSystemProducer.NullProducerMsg)
```

Comment:
Need sendFailed.inc?

Reply:
I thought about doing that, but according to the doc in KafkaSystemProducerMetrics, this metric is tracking failures in the KafkaProducer as opposed to "sends" which is tracking calls to the KafkaSystemProducer. I didn't want to change the semantics without good reason. 

Thoughts?